File: BlaiseMeta/BlaiseGraphics/src/main/java/com/googlecode/blaisemath/graphics/swing/MultilineTextRenderer.java
/*
 * MultilineTextRenderer.java
 * Created on Jan 2, 2013
 */

package com.googlecode.blaisemath.graphics.swing;

/*
 * #%L
 * BlaiseGraphics
 * --
 * Copyright (C) 2009 - 2014 Elisha Peterson
 * --
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */


import com.google.common.base.Strings;
import com.google.common.collect.Lists;
import static com.googlecode.blaisemath.graphics.swing.TextRenderer.anchorShift;
import com.googlecode.blaisemath.style.Anchor;
import com.googlecode.blaisemath.style.AttributeSet;
import com.googlecode.blaisemath.style.Renderer;
import com.googlecode.blaisemath.style.Styles;
import com.googlecode.blaisemath.util.AnchoredText;
import java.awt.Font;
import java.awt.Graphics2D;
import java.awt.Point;
import java.awt.font.FontRenderContext;
import java.awt.font.TextLayout;
import java.awt.geom.Point2D;
import java.awt.geom.Rectangle2D;
import java.util.Arrays;
import java.util.logging.Level;
import java.util.logging.Logger;

/**
 * Draws text with line breaks, respecting the location of the line breaks.
 *
 * @todo calculate boundaries of multiline text
 * @todo use text alignment
 * 
 * @author petereb1
 */
public class MultilineTextRenderer implements Renderer<AnchoredText, Graphics2D> {
    
    @Override
    public String toString() {
        return "MultilineTextRenderer";
    }

    public boolean contains(AnchoredText primitive, AttributeSet style, Point2D point) {
        return boundingBox(primitive, style).contains(point);
    }

    public boolean intersects(AnchoredText primitive, AttributeSet style, Rectangle2D rect) {
        return boundingBox(primitive, style).intersects(rect);
    }

    @Override
    public void render(AnchoredText text, AttributeSet style, Graphics2D canvas) {
        if (Strings.isNullOrEmpty(text.getText())) {
            return;
        }
        Font f = Styles.getFont(style);       
        canvas.setFont(f);
        canvas.setColor(style.getColor(Styles.FILL));
        
        double lineHeight = canvas.getFontMetrics().getHeight();
        Rectangle2D bounds = boundingBox(text, style);
        String[] lns = text.getText().split("\n|\r\n");
        double y0 = bounds.getMaxY();
        double x0 = bounds.getMinX();
        for (String s : Lists.reverse(Arrays.asList(lns))) {
            canvas.drawString(s, (float)x0, (float) y0);
            y0 -= lineHeight;
        }
    }

    public Rectangle2D boundingBox(AnchoredText text, AttributeSet style) {
        if (Strings.isNullOrEmpty(text.getText())) {
            return null;
        }
        
        Object anchor = style.get(Styles.TEXT_ANCHOR);
        if (!(anchor == null || anchor instanceof String || anchor instanceof Anchor)) {
            Logger.getLogger(TextRenderer.class.getName()).log(Level.WARNING,
                    "Invalid text anchor: {0}", anchor);
        }

        Font font = Styles.getFont(style);
        FontRenderContext frc = new FontRenderContext(font.getTransform(), true, true);
        
        double width = 0;
        double height = 0;
        double leading = 0;
        for (String l : text.getText().split("\n|\r\n")) {
            TextLayout tl = new TextLayout(text.getText(), font, frc);
            width = Math.max(width, font.getStringBounds(l, frc).getWidth());
            height += tl.getAscent()+tl.getDescent()+tl.getLeading();
            leading = tl.getLeading();
        }
        height -= leading;
        
        Anchor textAnchor = anchor == null ? Anchor.SOUTHWEST 
                : anchor instanceof Anchor ? (Anchor) anchor
                : anchor instanceof String ? Anchor.valueOf((String) anchor)
                : null;
        Point2D offset = style.getPoint(Styles.OFFSET, new Point());
        if (textAnchor == Anchor.SOUTHWEST) {
            return new Rectangle2D.Double(
                    text.getX() + offset.getX(), 
                    text.getY() + offset.getY()-height, 
                    width, height);
        }

        Point2D.Double shift = anchorShift(textAnchor, width, height);

        return new Rectangle2D.Double(
                text.getX() + offset.getX() + shift.x, 
                text.getY() + offset.getY() + shift.y-height, 
                width, height);
    }
    
}

File: blaise-graphics/src/main/java/com/googlecode/blaisemath/graphics/swing/render/MultilineTextRenderer.kt
package com.googlecode.blaisemath.graphics.swing.render

import com.google.common.base.Strings
import com.google.common.collect.Lists
import com.googlecode.blaisemath.graphics.Renderer
import com.googlecode.blaisemath.graphics.svg.SvgPathCoderTest
import com.googlecode.blaisemath.graphics.swing.ClippedImageTestFrame
import com.googlecode.blaisemath.graphics.swing.ContextMenuTestFrame
import com.googlecode.blaisemath.graphics.swing.HelloWorldTestFrame
import com.googlecode.blaisemath.graphics.swing.PanAndZoomTestFrame
import com.googlecode.blaisemath.graphics.swing.SelectionTestFrame
import com.googlecode.blaisemath.graphics.swing.TooltipTestFrame
import com.googlecode.blaisemath.primitive.Anchor
import com.googlecode.blaisemath.primitive.AnchoredText
import com.googlecode.blaisemath.style.AttributeSet
import com.googlecode.blaisemath.style.Styles
import junit.framework.TestCase
import java.awt.Graphics2D
import java.awt.Point
import java.awt.RenderingHints
import java.awt.Toolkit
import java.awt.font.FontRenderContext
import java.awt.geom.Point2D
import java.awt.geom.Rectangle2D
import java.util.*

/*
* #%L
* BlaiseGraphics
* --
* Copyright (C) 2009 - 2021 Elisha Peterson
* --
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*      http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
* #L%
*/ /**
 * Draw text on multiple lines, using line breaks provided with the text. By default, the text is anchored at the upper
 * left, so that text is drawn to the right and below the anchor point. For alternate anchors, all lines of text are
 * positioned in the same way, and the text may be centered, left-aligned, or right-aligned, depending on the anchor.
 *
 * @author Elisha Peterson
 */
class MultilineTextRenderer : Renderer<AnchoredText?, Graphics2D?> {
    override fun toString(): String {
        return "MultilineTextRenderer"
    }

    override fun contains(point: Point2D?, primitive: AnchoredText?, style: AttributeSet?, canvas: Graphics2D?): Boolean {
        return boundingBox(primitive, style, canvas).contains(point)
    }

    override fun intersects(rect: Rectangle2D?, primitive: AnchoredText?, style: AttributeSet?, canvas: Graphics2D?): Boolean {
        return boundingBox(primitive, style, canvas).intersects(rect)
    }

    override fun render(text: AnchoredText?, style: AttributeSet?, canvas: Graphics2D?) {
        if (Strings.isNullOrEmpty(text.getText())) {
            return
        }
        val font = Styles.fontOf(style)
        canvas.setFont(font)
        canvas.setColor(style.getColor(Styles.FILL))
        canvas.setRenderingHint(RenderingHints.KEY_FRACTIONALMETRICS, RenderingHints.VALUE_FRACTIONALMETRICS_ON)
        val frc = canvas.getFontRenderContext()
        val textAnchor = Styles.anchorOf(style, Anchor.SOUTHWEST)
        val lineHeight = font.getLineMetrics("", frc).height.toDouble()
        val bounds = boundingBox(text, style, canvas)
        val offset = style.getPoint2D(Styles.OFFSET, Point())!!
        val x0 = bounds.getMinX()
        var y0 = bounds.getMaxY()
        for (line in Lists.reverse(Arrays.asList(*lines(text)))) {
            val wid = font.getStringBounds(line, frc).width
            val dx = (textAnchor.offsetForRectangle(bounds.getWidth() - wid, 0.0).x
                    + 0.5 * (bounds.getWidth() - wid))
            canvas.drawString(line, (x0 + dx) as Float, y0 as Float)
            y0 -= lineHeight
        }
    }

    override fun boundingBox(text: AnchoredText?, style: AttributeSet?, canvas: Graphics2D?): Rectangle2D? {
        if (Strings.isNullOrEmpty(text.getText())) {
            return null
        }
        val font = Styles.fontOf(style)
        canvas?.setRenderingHint(RenderingHints.KEY_FRACTIONALMETRICS, RenderingHints.VALUE_FRACTIONALMETRICS_ON)
        val frc = if (canvas == null) FontRenderContext(font.transform, true, false) else canvas.fontRenderContext
        var width = 0.0
        val lines = lines(text)
        for (line in lines) {
            width = Math.max(width, font.getStringBounds(line, frc).width)
        }
        val lineHeight = font.getLineMetrics("", frc).height.toDouble()
        var height = lineHeight * lines.size
        height -= lineHeight - font.size * 72.0 / Toolkit.getDefaultToolkit().screenResolution
        val textAnchor = Styles.anchorOf(style, Anchor.NORTHWEST)
        val offset = style.getPoint2D(Styles.OFFSET, Point())!!
        return textAnchor.rectangleAnchoredAt(text.getX() + offset.x, text.getY() + offset.y,
                width, height)
    }

    companion object {
        private val INST: MultilineTextRenderer? = MultilineTextRenderer()
        fun getInstance(): Renderer<AnchoredText?, Graphics2D?>? {
            return INST
        }

        private fun lines(text: AnchoredText?): Array<String?>? {
            return text.getText().split("\n|\r\n".toRegex()).toTypedArray()
        }
    }
}